viam_rust_utils.h

```diff
@@ -0,0 +1,1222 @@
+ * For a rotation based on quaternions, see [`UnitQuaternion`](crate::UnitQuaternion) instead.
+ *
+ * Note that instead of using the [`Rotation`](crate::Rotation) type in your code directly, you should use one
+ * of its aliases: [`Rotation2`](crate::Rotation2), or [`Rotation3`](crate::Rotation3). Though
+ * keep in mind that all the documentation of all the methods of these aliases will also appears on
+ * this page.
+ *
+ * # Construction
+ * * [Identity <span style="float:right;">`identity`</span>](#identity)
+ * * [From a 2D rotation angle <span style="float:right;">`new`…</span>](#construction-from-a-2d-rotation-angle)
+ * * [From an existing 2D matrix or rotations <span style="float:right;">`from_matrix`, `rotation_between`, `powf`…</span>](#construction-from-an-existing-2d-matrix-or-rotations)
+ * * [From a 3D axis and/or angles <span style="float:right;">`new`, `from_euler_angles`, `from_axis_angle`…</span>](#construction-from-a-3d-axis-andor-angles)
+ * * [From a 3D eye position and target point <span style="float:right;">`look_at`, `look_at_lh`, `rotation_between`…</span>](#construction-from-a-3d-eye-position-and-target-point)
+ * * [From an existing 3D matrix or rotations <span style="float:right;">`from_matrix`, `rotation_between`, `powf`…</span>](#construction-from-an-existing-3d-matrix-or-rotations)
+ *
+ * # Transformation and composition
+ * Note that transforming vectors and points can be done by multiplication, e.g., `rotation * point`.
+ * Composing an rotation with another transformation can also be done by multiplication or division.
+ * * [3D axis and angle extraction <span style="float:right;">`angle`, `euler_angles`, `scaled_axis`, `angle_to`…</span>](#3d-axis-and-angle-extraction)
+ * * [2D angle extraction <span style="float:right;">`angle`, `angle_to`…</span>](#2d-angle-extraction)
+ * * [Transformation of a vector or a point <span style="float:right;">`transform_vector`, `inverse_transform_point`…</span>](#transformation-of-a-vector-or-a-point)
+ * * [Transposition and inversion <span style="float:right;">`transpose`, `inverse`…</span>](#transposition-and-inversion)
+ * * [Interpolation <span style="float:right;">`slerp`…</span>](#interpolation)
+ *
+ * # Conversion
+ * * [Conversion to a matrix <span style="float:right;">`matrix`, `to_homogeneous`…</span>](#conversion-to-a-matrix)
+ *
+ */
+typedef struct Rotation_f64__3 {
+  SMatrix_f64__3__3 matrix;
+} Rotation_f64__3;
+
+/**
+ * A 3-dimensional rotation matrix.
+ *
+ * **Because this is an alias, not all its methods are listed here. See the [`Rotation`](crate::Rotation) type too.**
+ */
+typedef struct Rotation_f64__3 Rotation3_f64;
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+/**
+ * Initialize a tokio runtime to run a gRPC client/sever, user should call this function before trying to dial to a Robot
+ * Returns a pointer to a [`DialFfi`]
+ */
+struct DialFfi *init_rust_runtime(void);
```

Comment:
Is there an option to have bindgen prefix these with a leading string to act as a namespace? So they come out as `viam_rust_utils_init_rust_runtime` or similar?

Yes, they are prolix, but `dial` and `free_string` and `quaternion_add` are very general names with a high chance of very annoying collisions.

Reply:
unfortunately no, there is a prefix option but it prefixes the whole line so we'd end up with `viam_rust_utils_ struct DialFfi *init_rust_runtime(void);` which obviously doesn't compile.

I think the relatively straightforward fix is to just rename the methods on the rust side. I do want to think through how important this is, however, because I think this would be an extremely breaking change.

Reply:
So, it can't be a breaking change to do the renames in this header. I think you can use the same cbindgen renaming tricks to achieve that, and I think it is worth it.

Simultaneously, I also think the rust FFI functions should be renamed, per https://viam.atlassian.net/browse/RSDK-1978. The way to do that is probably to rename the functions, add forwarding functions under the old names, and attach deprecation attributes to the old names.

But I'd say we can start by correcting them here, and that'll inform the choices for 1978. Alternatively, just fold 1978 into this?

Reply:
Done, opted to fold 1978 into this. I opted to prefix our types and functions with `viam_` rather than `viam_rust_utils_`. My reasoning was 1) this is a lot less verbose, and 2) the viam_ prefix probably doesn't create a significantly higher risk of collision than the viam_rust_utils_ prefix. Happy to revisit if either of you disagree with point 2, or think there's some other reason to prefer the fuller prefix.